api/sendfild/service/waitmanager/waitrunner.go

```diff
@@ -0,0 +1,188 @@
+	w.waitAndNotify()
+}
+
+type CancelListenerFunc = func()
+
+func (w *WaitRunner) AddListener(listener chan WaitResult) CancelListenerFunc {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	w.listeners = append(w.listeners, listener)
+	return func() {
+		w.lock.Lock()
+		defer w.lock.Unlock()
+		index := -1
+		for i, l := range w.listeners {
+			if l == listener {
+				index = i
+				break
+			}
+		}
+		if index >= 0 {
+			// ToDo: Write unit test to make sure adding multiple listeners and removing some works for remaining listeners.
+			w.listeners[index] = w.listeners[len(w.listeners)-1]
+			w.listeners = w.listeners[:len(w.listeners)-1]
+		}
+	}
+}
+
+func (w *WaitRunner) Close() error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	w.closeLotusClient()
+	w.waitCtxCancel()
+	w.mainCtxCancel()
+	return nil
+}
+
+func (w *WaitRunner) waitAndNotify() {
+	go func() {
+		c, err := cid.Decode(w.messageCid)
+		if err != nil {
+			if err := w.store.FailTxn(w.mainCtx, w.messageCid, err.Error()); err != nil {
+				w.notifyErr(fmt.Errorf("failing txn in store: %v", err))
+				return
+			}
+			w.notify(w.messageCid)
+			return
+		}
+
+		if err := w.store.SetWaiting(w.mainCtx, w.messageCid, true); err != nil {
+			w.notifyErr(fmt.Errorf("setting txn to waiting: %v", err))
+			return
+		}
+
+		res, err := w.lotusClient.StateWaitMsg(w.waitCtx, c, w.confidence)
+
+		if err := w.store.SetWaiting(w.mainCtx, w.messageCid, false); err != nil {
+			w.notifyErr(fmt.Errorf("setting txn to not waiting: %v", err))
+			return
+		}
+
+		if err != nil {
+			// If for some reason the lotus node doesn't know about the cid, consider that a final error.
+			if strings.Contains(err.Error(), "block not found") {
+				log.Errorf("calling StateWaitMsg block not found: %s", err.Error())
+				if err := w.store.FailTxn(w.mainCtx, w.messageCid, "block not found in lotus"); err != nil {
+					w.notifyErr(fmt.Errorf("failing txn in store: %v", err))
+					return
+				}
+				w.notify(w.messageCid)
+				return
+			}
+			// ugly but seems to be the only way to detect this
+			if err.Error() == "context canceled" {
+				w.notifyErr(fmt.Errorf("waiting for txn status timed out, but txn is still processing, query txn again if needed"))
+				return
+			}
+			w.notifyErr(fmt.Errorf("calling StateWaitMsg: %v", err))
+			return
+		}
+
+		if res.Receipt.ExitCode.IsError() {
+			if res.Receipt.ExitCode.IsSendFailure() {
+				log.Errorf("received exit code send failure: %s", res.Receipt.ExitCode.String())
+			} else {
+				log.Infof("received exit code error: %s", res.Receipt.ExitCode.String())
+			}
+			if err := w.store.FailTxn(w.mainCtx, w.messageCid, fmt.Sprintf("error exit code: %v", res.Receipt.ExitCode.Error())); err != nil {
+				w.notifyErr(fmt.Errorf("failing txn in store: %v", err))
+				return
+			}
+
+			cidToNotify := c.String()
+			if res.Message.Defined() {
+				cidToNotify = res.Message.String()
+			}
+
+			w.notify(cidToNotify)
+			return
```

Comment:
Depending on the type of receipt exit code error, I wasn't sure if some cases would still result in a message cid, so just doing this check just in case and calling back to the known message cid our runner was created with.